node/router/src/helpers/peer.rs

```diff
@@ -13,108 +13,127 @@
-    first_seen: Instant,
+    pub first_seen: Instant,
     /// The timestamp of the last message received from this peer.
-    last_seen: Instant,
+    pub last_seen: Instant,
+    /// A reference to the associated `Router` object.
+    pub router: Router<N>,
 }
 
 impl<N: Network> Peer<N> {
-    /// Initializes a new instance of `Peer`.
-    pub fn new(listening_ip: SocketAddr, connected_ip: SocketAddr, challenge_request: &ChallengeRequest<N>) -> Self {
-        Self {
-            peer_ip: listening_ip,
-            peer_addr: connected_ip,
-            address: challenge_request.address,
-            node_type: challenge_request.node_type,
-            version: challenge_request.version,
-            first_seen: Instant::now(),
-            last_seen: Instant::now(),
-        }
-    }
-
-    /// Returns the IP address of the peer, with the port set to the listener port.
-    pub const fn ip(&self) -> SocketAddr {
-        self.peer_ip
-    }
-
-    /// Returns the connected address of the peer.
-    pub const fn addr(&self) -> SocketAddr {
-        self.peer_addr
+    /// Create a candidate peer.
+    pub const fn new_candidate(listener_addr: SocketAddr) -> Self {
+        Self::Candidate(CandidatePeer { listener_addr, restricted: None })
     }
 
-    /// Returns the Aleo address of the peer.
-    pub const fn address(&self) -> Address<N> {
-        self.address
+    /// Promote a connecting peer to a fully connected one.
+    pub fn upgrade_to_connected(&mut self, connected_addr: SocketAddr, cr: &ChallengeRequest<N>, router: Router<N>) {
+        // Logic check: this can only happen during the handshake.
+        assert!(matches!(self, Self::Connecting));
+
+        let timestamp = Instant::now();
+        let listener_addr = SocketAddr::from((connected_addr.ip(), cr.listener_port));
+
+        // Introduce the peer in the resolver.
+        router.resolver.write().insert_peer(listener_addr, connected_addr);
+
+        *self = Self::Connected(ConnectedPeer {
+            listener_addr,
+            connected_addr,
+            aleo_addr: cr.address,
+            node_type: cr.node_type,
+            version: cr.version,
+            first_seen: timestamp,
+            last_seen: timestamp,
+            router,
+        });
     }
 
-    /// Returns the node type.
-    pub const fn node_type(&self) -> NodeType {
-        self.node_type
-    }
-
-    /// Returns `true` if the peer is a validator.
-    pub const fn is_validator(&self) -> bool {
-        self.node_type.is_validator()
-    }
-
-    /// Returns `true` if the peer is a prover.
-    pub const fn is_prover(&self) -> bool {
-        self.node_type.is_prover()
-    }
-
-    /// Returns `true` if the peer is a client.
-    pub const fn is_client(&self) -> bool {
-        self.node_type.is_client()
-    }
+    /// Demote a peer to candidate status, marking it as disconnected.
+    pub fn downgrade_to_candidate(&mut self, listener_addr: SocketAddr, restrict: bool) {
```

Comment:
This is great! Can we add unit tests for downgrading peers, or do they exist already somewhere?

Reply:
Indirectly yes, as we have tests to detect the number of connected peers. It's not really testable in a truly isolated unit fashion, as it involves the resolver too.